src/backend/distributed/operations/repair_shards.c

```diff
@@ -287,7 +429,120 @@ citus_move_shard_placement(PG_FUNCTION_ARGS)
 
 
 /*
- * master_move_shard_placement is a wrapper function for old UDF name.
+ * EnsureEnoughDiskSpaceForShardMove checks that there is enough space for
+ * shard moves of the given colocated shard list from source node to target node.
+ * It tries to clean up old shard placements to ensure there is enough space.
+ */
+static void
+EnsureEnoughDiskSpaceForShardMove(List *colocatedShardList,
+								  char *sourceNodeName, uint32 sourceNodePort,
+								  char *targetNodeName, uint32 targetNodePort)
+{
+	int64 colocationSizeInBytes = ColocationSizeInBytes(colocatedShardList,
+														sourceNodeName,
+														sourceNodePort);
+
+	uint32 connectionFlag = 0;
+	MultiConnection *connection = GetNodeConnection(connectionFlag, targetNodeName,
+													targetNodePort);
+	int checkSpaceTries = 0;
+	while (true)
+	{
+		bool droppedAllOldShards = false;
+		if (checkSpaceTries > 0)
+		{
+			/*
+			 * Only run shard cleanup when the first round of checks has failed
+			 * otherwise simply let the automatic cleanup do its job.
+			 *
+			 * NOTE: this tries to drop all marked shards, not only on the shard
+			 * that we're trying to move to.
+			 */
+			ereport(NOTICE, (errmsg("trying to drop shards to resolve issue...")));
+
+			bool waitForCleanupLock = true;
+			droppedAllOldShards = DropMarkedShards(waitForCleanupLock, NULL);
+
+			/*
+			 * NOTE: It actually doesn't make sense to wait in some cases where
+			 * it does droppedAllOldShards is false. Most importantly the case
+			 * where there are no shards marked for removal on the target
+			 * shard. It's not a big issue though, since in this case we simply
+			 * error out after a couple of more rounds of trying to drop the
+			 * left over shards.
+			 */
+		}
+
+		DeferredErrorMessage *temporaryError = CheckSpaceConstraints(connection,
+																	 colocationSizeInBytes);
+		if (temporaryError == NULL)
+		{
+			break;
+		}
+
+		if (droppedAllOldShards)
+		{
+			/*
+			 * If we dropped all shards but still don't have enough space for shard move
+			 * then we don't have any choice but error
+			 */
+			RaiseDeferredError(temporaryError, ERROR);
+		}
+
+		RaiseDeferredError(temporaryError, WARNING);
+		checkSpaceTries++;
+		if (checkSpaceTries >= WaitForDeferShardsMaxTries)
+		{
+			ereport(ERROR, (
+						errmsg(
+							"shards were not removed after trying %d times, giving up",
+							checkSpaceTries),
+						errhint("Try increasing citus.move_shard_defer_drop_tries")
+						));
+		}
+
+		/* wait the timeout and try again*/
+		WaitForMiliseconds(WaitForDeferShardRetryTimeInSec * 1000);
+	}
+}
+
+
+/*
+ * citus_disk_available returns the number of bytes of available diskspace that
+ * are on the given tablespace.
+ */
+Datum
+citus_disk_available(PG_FUNCTION_ARGS)
+{
+	struct statvfs diskstats;
+	int error = statvfs("base/", &diskstats);
```

Comment:
fwiw, this PR also adds some similar functions:

https://github.com/citusdata/citus/pull/4383/files#diff-a0fe2d7cbca8cd88f0580ff39c74be56eb992920a2b2acd1405e71cf4095a0bfR239

Reply:
I think the API Marco noted looks better, a single UDF call gets both of the disk stats at once.

Also, I have not dived on the internals of the functions, but @SaitTalhaNisanci can you make sure that if they are the same? If not, can you asses which one is more accurate?

Reply:
They seem quite similar, I will go with the single UDF version. Also what these fields mean:

![image](https://user-images.githubusercontent.com/17179875/116979506-bfbc6480-accd-11eb-89e7-84c2b5fe6bad.png)

f_bsize and f_frsize seems to be the same almost all the time so I guess there isn't much difference in using either